content/dakshdeephere/threading.md

```diff
@@ -0,0 +1,355 @@
+
+## Multithreading
+
+- A `thread` is a path which is followed during a program’s execution.
+- Let's consider an instance where a program cannot simultaneously read keystrokes and create drawings. These activities pose a challenge for the program as they cannot be performed concurrently. However, this issue can be resolved through multitasking, enabling the program to execute multiple tasks simultaneously.
+- Multitasking is of two types: `Processor based` and `thread based`.
+- `Processor based` multitasking is totally managed by the OS, however multitasking through multithreading can be controlled by the programmer to some extent.
+- The concept of multi-threading needs proper understanding of these two terms – a `process` and a `thread`.
+- A process is a program being executed. A process can be further divided into independent units known as threads.
+- A `thread` is like a small light-weight process within a process Or we can say a collection of threads is what is known as a process.
+
+## Process vs Thread
+
+| S.N.      | Process | Thread |
+| -- | ----------- | ----------- |
+| 1      | Process is heavy weight or resource intensive.| Thread is light weight, taking lesser resources than a process.      |
+| 2   | Process switching needs interaction with operating system.        | Thread switching does not need to interact with operating system. |
+| 3 | In multiple processing environments, each process executes the same code but has its own memory and file resources. | All threads can share same set of open files, child processes. |
+| 4 | If one process is blocked, then no other process can execute until the first process is unblocked. | While one thread is blocked and waiting, a second thread in the same task can run. |
+| 5 | Multiple processes without using threads use more resources. | Multiple threaded processes use fewer resources. |
+| 6 | In multiple processes each process operates independently of the others. | One thread can read, write or change another thread's data. |
+
+## Challenges for Programmers while creating Threads
+
+- `Dividing activities` – It involves finding the functions within the job that can be run in parallel on separate processors.
+- `Balance` – The task assigned to each processor must also be equal. Now there can be different parameters for that. One parameter can be, assign equal tasks to each processor. But, tasks assigned to more processor may require higher execution time thus overloading one processor. Thus, simply assigning equal tasks to each processor may not work.
+- `Data splitting` – Another challenge is to split the data required for each task.
+- `Data dependency` – sometimes the data required by one thread (T1) might be produced by another (T2). Thus, T1 can not run before T2. Therefore, it becomes difficult for programmers to code.
+- `Testing and debugging` – Multiple threads running in parallel on multiple cores poses another challenge in the testing of applications.
+
+## Syntax
```

Comment:
```suggestion
## Creating a thread
```

Reply:
Let's turn the content below into a step-by-step tutorial and slowly add in code w/ explanations